design/service-apis.md

```diff
@@ -0,0 +1,384 @@
+}
+```
+
+The `gatewayClassRef` field is a reference to a GatewayClass used for managing the Contour. If unspecified, the field
+will default to "None", indicating a GatewayClass does not exist, and the operator should use the Contour resource to
+instantiate a Contour instance.
+
+The following is an example of a Contour resource that references a GatewayClass:
+```yaml
+apiVersion: operator.projectcontour.io/v1alpha1
+kind: Contour
+metadata:
+  name: example-contour
+  namespace: contour-operator
+spec:
+  gatewayClassRef: example-gatewayclass
+  replicas: 2
+  namespace:
+    name: projectcontour
+```
+
+### Gateway
+
+A Gateway describes how traffic can be translated to Services within a Kubernetes cluster. That is, it defines how to
+translate traffic from somewhere that does not know about Kubernetes to somewhere that does. For example, traffic sent
+from a client outside the cluster to a Service resource through a proxy such as Contour. While many use-cases have
+client traffic originating “outside” the cluster, this is not a requirement.
+
+A Gateway is 1:1 with the lifecycle of the infrastructure configuration. When a Gateway is created, the operator will
+create an instance of Contour. Gateway is the resource that triggers actions in Service APIs. Other Service API
+resources represent structured configuration that are linked together by a Gateway.
+
+A Gateway MAY contain one or more Service APIs *Route references which serve to forward traffic to a specific service
+based on application-level rules. Note that Contour Operator is NOT responsible for managing *Route Service API
+resources.
+
+The following is a Gateway example that listens for HTTP traffic on TCP port 80:
+
+```yaml
+kind: Gateway
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gateway
+  namespace: projectcontour
+spec:
+  gatewayClassName: example-gatewayclass
+  listeners:
+    - protocol: HTTP
+      port: 80
+  ...
+```
+
+The `port` or `protocol` fields are required and specify the network port/protocol pair that Envoy will listen on for
+connection requests. The operator will use the value of `port` for the Envoy Service resource. If multiple listeners are
+specified with unique ports, the operator will represent each port in the Envoy Service. The above Gateway example will
+create the following Envoy Service:
+
+```yaml
+kind: Service
+apiVersion: v1
+metadata:
+  name: envoy
+  namespace: projectcontour
+  ...
+spec:
+  ports:
+  - port: 80
+    name: http
+    protocol: TCP
+   ...
+```
+
+__Note:__ Contour is responsible for constructing routing rules and forwarding traffic based on other `listeners`
+fields.
+
+TLS configuration must be included in the Gateway listener when "HTTPS" or "TLS" is specified for
+`protocol`. The `certificateRef` field is a reference to a Secret in the Gateway's namespace containing a TLS
+certificate and private key. The secret MUST contain the following keys and data:
```

Comment:
Maybe a note that the `kubernetes.io/tls` Secret type has these?

Reply:
Good point, I'll update the doc and xref: https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets